File: Session5_DFS/validateBST.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */

class Solution {
    /**
     * validateHelper is a recursive function that traverses the
     * tree DFS-style and records the Lower Bound and Upper Bound 
     * of whatever the node shoulld be.
     */
    public boolean validateHelper(TreeNode node, Integer min, Integer max) {
        /**
         * Tried using Integer.MAX_VALUE to no avail. Some Int type
         * problem, so I chose to use null instead. Anyhow, empty node
         * does not influence whether BST is valid or not
         */  
        if (node == null)
            return true;

        /**
         * If bounds are defined and node value lies outside these 
         * bounds, then return false
         */
        if ((min != null && node.val <= min) || (max != null && node.val >= max))
            return false;

        /* To be a valid BST, all nodes must be valid. */
        return (validateHelper(node.left, min, node.val) && validateHelper(node.right, node.val, max));
    }

    public boolean isValidBST(TreeNode root) {
        /* Initialize recursion with null bounds (root always valid) */
        return validateHelper(root, null, null);
    }
}

File: Session5_DFS/validateBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def validateHelper(self, node: TreeNode, lowBound, highBound) -> bool:
        # Empty node does not influence whether BST is Valid
        if not node:
            return True
        
        # If Bounds are defined and node value lies outside these
        # bounds, then return false
        if (lowBound != None and node.val <= lowBound or 
            highBound != None and node.val >= highBound):
            return False
        
        """
        In a sense, lowBound and highBound are accumulators storing the 
        highest and lowest value seen so far during the traversal.
        """
        # Lower bound remains. However, left child node should be LOWER than parent value
        validateLeft = self.validateHelper(node.left, lowBound, node.val)
        # Upper bound remains. However, right child node should be HIGHER than parent value
        validateRight = self.validateHelper(node.right, node.val, highBound)
        
        # To be a valid BST, all nodes must be valid.
        return validateLeft and validateRight
    
    def isValidBST(self, root: TreeNode) -> bool:
        # Could also put -infinity and infintiy here to start the recursion
        # Root node is always a valid BST (No bounds to compare to)
        return self.validateHelper(root, None, None)